### PR TITLE
Fix uniform neighborhood sampling remove duplicates

### LINKED ISSUE #2226: Refactor libcugraph Neighborhood Sampling
The following steps are required (each a separate PR):
1. Update API as discussed in our meeting (remove batching, drop some of the return values)
2. Refactor the implementation to use more of the existing primitives and support both SG and MG

### PR DESCRIPTION
**Pull request by ChuckHastings**
@jnke2016 observed that the recent sampling code can return duplicates.

After some discussion it was decided to remove the duplicates and return a count of how many duplicates were removed.  The C++ implementation is updated in this PR to do that.  The C API will remain the same but is ready to support it once we have stopped using the old implementation.

Closes #2226

### CODE REVIEW

File: cpp/include/cugraph/algorithms.hpp [hunk 1]
```diff
@@ -1587,22 +1587,25 @@ uniform_nbr_sample(raft::handle_t const& handle,
  * @param with_replacement boolean flag specifying if random sampling is done with replacement
  * (true); or, without replacement (false); default = true;
  * @param seed A seed to initialize the random number generator
- * @return tuple device vectors (vertex_t source_vertex, vertex_t destination_vertex, weight_t wgt)
+ * @return tuple device vectors (vertex_t source_vertex, vertex_t destination_vertex, weight_t wgt,
+ * edge_t count)
```
Comment (seunghwak) [inline cpp/include/cugraph/algorithms.hpp:1591]: `wgt` => `weight` better not use abbreviations in the documentation.
And what is `edge_t count`? # of duplicates? I think this needs more explanation.

File: cpp/src/sampling/detail/sampling_utils_impl.cuh [hunk 2]
```diff
@@ -907,5 +907,39 @@ gather_one_hop_edgelist(
...
+                                        count.begin(),
+                                        tuple_iter,
+                                        count.begin());
+
+  size_t sz = thrust::distance(tuple_iter, end_iter.first);
+
+  src.resize(sz, handle.get_stream());
+  dst.resize(sz, handle.get_stream());
+  wgt.resize(sz, handle.get_stream());
+  count.resize(sz, handle.get_stream());
```
Comment (seunghwak) [inline cpp/src/sampling/detail/sampling_utils_impl.cuh:939]: Call `shrink_to_fit` to free-up memory.

And even better is we can count the number of unique elements and just allocate the exact amount of memory (so no need to resize).

See
```
    auto num_uniques = thrust::count_if(
      rmm::exec_policy(stream_view),
      thrust::make_counting_iterator(size_t{0}),
      thrust::make_counting_iterator(edgelist_majors.size()),
      detail::is_first_in_run_pair_t<vertex_t>{edgelist_majors.data(), edgelist_minors.data()});

    rmm::device_uvector<vertex_t> tmp_edgelist_majors(num_uniques, stream_view);
    rmm::device_uvector<vertex_t> tmp_edgelist_minors(tmp_edgelist_majors.size(), stream_view);
    rmm::device_uvector<weight_t> tmp_edgelist_weights(tmp_edgelist_majors.size(), stream_view);
    thrust::reduce_by_key(rmm::exec_policy(stream_view),
                          pair_first,
                          pair_first + edgelist_majors.size(),
                          (*edgelist_weights).begin(),
                          thrust::make_zip_iterator(thrust::make_tuple(
                            tmp_edgelist_majors.begin(), tmp_edgelist_minors.begin())),
                          tmp_edgelist_weights.begin());
```
https://github.com/rapidsai/cugraph/blob/branch-22.06/cpp/src/structure/coarsen_graph_impl.cuh#L82

File: cpp/include/cugraph/algorithms.hpp (hunk 1 above)
Comment (ChuckHastings) [reply cpp/include/cugraph/algorithms.hpp:1591]: Updated documentation in next push.

File: cpp/src/sampling/detail/sampling_utils_impl.cuh (hunk 2 above)
Comment (ChuckHastings) [reply cpp/src/sampling/detail/sampling_utils_impl.cuh:939]: Change made in next push.

**Review by seunghwak — APPROVED** [review]

**Comment by BradReesWork** [comment]
@gpucibot merge

**Pull request merged**